packages/headless/src/controllers/facets/_common/facet-id-generator.ts

```diff
@@ -0,0 +1,50 @@
+import {FacetType} from '../../../features/facets/facet-api/request';
+import {AnyFacetSetState} from '../../../features/facets/generic/interfaces/generic-facet-state';
+
+export interface Logger {
+  warn: (message: string) => void;
+}
+
+export interface FacetIdConfig {
+  type: FacetType;
+  field: string;
+  state: AnyFacetSetState;
+}
+
+export function generateFacetId(config: FacetIdConfig, logger: Logger) {
+  const {type, field, state} = config;
+
+  const prefix = `${type}_${field}_`;
+  const id = calculateId(prefix, state);
+
+  logWarningMessageIfNeeded(id, field, logger);
+
+  return `${prefix}${id}`;
+}
+
+function calculateId(prefix: string, state: AnyFacetSetState) {
+  const keys = Object.keys(state);
+  return findMaxId(keys, prefix) + 1;
+}
+
+function findMaxId(keys: string[], prefix: string) {
+  const ids = keys.map((key) => {
```

Comment:
The client has the option to set the id manually and if not we generate one right? And that's why you used `map` instead of `filter`?

Reply:
If the client specifies an id, then we will always use it, and the `generateFacetId` code will not run.

I used `map` here because I want to transform every string id to just the number at the end of the id. If do not find a number, then the `map` function returns `-1`

Reply:
What happens if a client specifies the same id on two different facet? the `generateFacetId` but we it log a warning?

Reply:
Nope, there is no defense against this atm. Will open a jira for this,